# CARGAR Dataframes

In [1]:
import pandas as pd

# Cargar el conjunto de datos de entrenamiento
dataframeTrain = pd.read_csv("credit-train.csv")

# Cargar el conjunto de datos de prueba
dataframeTest = pd.read_csv("credit-test.csv")


### Descripción general del dataframe

In [4]:
dataframeTrain.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41188 entries, 0 to 41187
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             41188 non-null  int64  
 1   job             41188 non-null  object 
 2   marital         41188 non-null  object 
 3   education       41188 non-null  object 
 4   default         41188 non-null  object 
 5   housing         41188 non-null  object 
 6   loan            41188 non-null  object 
 7   contact         41188 non-null  object 
 8   month           41188 non-null  object 
 9   day_of_week     41188 non-null  object 
 10  duration        41188 non-null  int64  
 11  campaign        41188 non-null  int64  
 12  pdays           41188 non-null  int64  
 13  previous        41188 non-null  int64  
 14  poutcome        41188 non-null  object 
 15  emp_var_rate    41188 non-null  float64
 16  cons_price_idx  41188 non-null  float64
 17  cons_conf_idx   41188 non-null 


El dataframe contiene 41,188 filas y 21 columnas.
Las columnas contienen datos de diferentes tipos: float64 (5 columnas), int64 (6 columnas) y object (10 columnas).
No hay valores nulos en el dataframe, lo que significa que todas las columnas tienen 41,188 valores no nulos.

In [6]:
print(dataframeTrain.head())
print(dataframeTest.head())

   age          job  marital          education  default housing loan  \
0   44  blue-collar  married           basic.4y  unknown     yes   no   
1   53   technician  married            unknown       no      no   no   
2   28   management   single  university.degree       no     yes   no   
3   39     services  married        high.school       no      no   no   
4   55      retired  married           basic.4y       no     yes   no   

    contact month day_of_week  ...  campaign  pdays  previous     poutcome  \
0  cellular   aug         thu  ...         1    999         0  nonexistent   
1  cellular   nov         fri  ...         1    999         0  nonexistent   
2  cellular   jun         thu  ...         3      6         2      success   
3  cellular   apr         fri  ...         2    999         0  nonexistent   
4  cellular   aug         fri  ...         1      3         1      success   

  emp_var_rate  cons_price_idx  cons_conf_idx  euribor3m  nr_employed  y  
0          1.4   

Estadísticas descriptivas para columnas numéricas:

La columna age tiene una media de aproximadamente 40 años, con un rango de 17 a 98 años.
La duración de la llamada (duration) varía desde 0 hasta 4918 segundos, con una media de aproximadamente 258 segundos.
El número de contactos durante esta campaña (campaign) oscila entre 1 y 56, con una media de aproximadamente 2.57 contactos.
La mayoría de los valores en la columna pdays son 999, lo que indica que la mayoría de los clientes no fueron contactados previamente.
La mayoría de los clientes no tienen valores previos (previous) o tienen solo unos pocos valores anteriores.
Las columnas emp_var_rate, cons_price_idx, cons_conf_idx, euribor3m y nr_employed contienen valores relacionados con indicadores económicos.

## ANALISIS Columnas Dataframes

In [4]:
print("Columnas del conjunto de datos de entrenamiento:")
print(dataframeTrain.columns)

print("\nColumnas del conjunto de datos de prueba:")
print(dataframeTest.columns)


Columnas del conjunto de datos de entrenamiento:
Index(['age', 'job', 'marital', 'education', 'default', 'housing', 'loan',
       'contact', 'month', 'day_of_week', 'duration', 'campaign', 'pdays',
       'previous', 'poutcome', 'emp_var_rate', 'cons_price_idx',
       'cons_conf_idx', 'euribor3m', 'nr_employed', 'y'],
      dtype='object')

Columnas del conjunto de datos de prueba:
Index(['age', 'job', 'marital', 'education', 'default', 'housing', 'loan',
       'contact', 'month', 'day_of_week', 'duration', 'campaign', 'pdays',
       'previous', 'poutcome', 'emp_var_rate', 'cons_price_idx',
       'cons_conf_idx', 'euribor3m', 'nr_employed'],
      dtype='object')


## ANALISIS 5 primers filas Dataframes

In [5]:
print(dataframeTrain.head())
print(dataframeTest.head())

   age          job  marital          education  default housing loan  \
0   44  blue-collar  married           basic.4y  unknown     yes   no   
1   53   technician  married            unknown       no      no   no   
2   28   management   single  university.degree       no     yes   no   
3   39     services  married        high.school       no      no   no   
4   55      retired  married           basic.4y       no     yes   no   

    contact month day_of_week  ...  campaign  pdays  previous     poutcome  \
0  cellular   aug         thu  ...         1    999         0  nonexistent   
1  cellular   nov         fri  ...         1    999         0  nonexistent   
2  cellular   jun         thu  ...         3      6         2      success   
3  cellular   apr         fri  ...         2    999         0  nonexistent   
4  cellular   aug         fri  ...         1      3         1      success   

  emp_var_rate  cons_price_idx  cons_conf_idx  euribor3m  nr_employed  y  
0          1.4   

In [6]:
dataframeTest.columns

Index(['age', 'job', 'marital', 'education', 'default', 'housing', 'loan',
       'contact', 'month', 'day_of_week', 'duration', 'campaign', 'pdays',
       'previous', 'poutcome', 'emp_var_rate', 'cons_price_idx',
       'cons_conf_idx', 'euribor3m', 'nr_employed'],
      dtype='object')

## Recuento de valores únicos en cada columna:

Las columnas categóricas tienen un rango diferente de valores únicos, lo que indica la variabilidad en las características de los cliente



In [7]:
dataframeTrain.nunique()

age                 78
job                 12
marital              4
education            8
default              3
housing              3
loan                 3
contact              2
month               10
day_of_week          5
duration          1544
campaign            42
pdays               27
previous             8
poutcome             3
emp_var_rate        10
cons_price_idx      26
cons_conf_idx       26
euribor3m          316
nr_employed         11
y                    2
dtype: int64

## Recuento de valores nulos en cada columna:.


In [8]:
dataframeTrain.isnull().sum()

age               0
job               0
marital           0
education         0
default           0
housing           0
loan              0
contact           0
month             0
day_of_week       0
duration          0
campaign          0
pdays             0
previous          0
poutcome          0
emp_var_rate      0
cons_price_idx    0
cons_conf_idx     0
euribor3m         0
nr_employed       0
y                 0
dtype: int64

# Creación de variables dummy:
Convertir las variables categóricas en variables binarias para que puedan ser utilizadas en modelos de machine learning.

In [8]:
# TRAIN Crear variables dummy para las columnas categóricas 
dataframeTrain = pd.get_dummies(dataframeTrain, columns=['job', 'marital', 'education', 'default', 'housing', 'loan', 'contact', 'month', 'day_of_week', 'poutcome'])

In [9]:
# TEST Crear variables dummy para las columnas categóricas 
dataframeTest = pd.get_dummies(dataframeTest, columns=['job', 'marital', 'education', 'default', 'housing', 'loan', 'contact', 'month', 'day_of_week', 'poutcome'])

## Normalización de datos:
Normalizar los datos numéricos puede mejorar el rendimiento de algunos algoritmos de machine learning.

In [10]:
#TRAIN
from sklearn.preprocessing import StandardScaler

# Inicializar el scaler
scaler = StandardScaler()

# Normalizar las columnas numéricas
dataframeTrain[['age', 'duration', 'campaign', 'pdays', 'previous', 'emp_var_rate', 'cons_price_idx', 'cons_conf_idx', 'euribor3m', 'nr_employed']] = scaler.fit_transform(dataframeTrain[['age', 'duration', 'campaign', 'pdays', 'previous', 'emp_var_rate', 'cons_price_idx', 'cons_conf_idx', 'euribor3m', 'nr_employed']])


In [11]:
#TEST
from sklearn.preprocessing import StandardScaler

# Inicializar el scaler
scaler = StandardScaler()

# Normalizar las columnas numéricas
dataframeTest[['age', 'duration', 'campaign', 'pdays', 'previous', 'emp_var_rate', 'cons_price_idx', 'cons_conf_idx', 'euribor3m', 'nr_employed']] = scaler.fit_transform(dataframeTest[['age', 'duration', 'campaign', 'pdays', 'previous', 'emp_var_rate', 'cons_price_idx', 'cons_conf_idx', 'euribor3m', 'nr_employed']])


## Eliminación de columnas no relevantes:
Eliminar columnas que no son relevantes para el modelo puede simplificar el análisis y mejorar la precisión del modelo.

In [12]:
dataframeTrain.drop(columns=['duration'], inplace=True)
dataframeTest.drop(columns=['duration'], inplace=True)

## Partición de datos:
Dividir el dataframe en conjunto de entrenamiento y conjunto de prueba para evaluar el rendimiento del modelo.
Código para dividir datos:

In [14]:
print(dataframeTrain.head())
print(dataframeTest.head())

        age  campaign     pdays  previous  emp_var_rate  cons_price_idx  \
0  0.381527 -0.565922  0.195414 -0.349494      0.839061       -0.227465   
1  1.245157 -0.565922  0.195414 -0.349494     -0.115781       -0.649003   
2 -1.153816  0.156105 -5.117342  3.691766     -1.134279        0.828107   
3 -0.098268 -0.204909  0.195414 -0.349494     -1.197935       -0.864955   
4  1.437075 -0.565922 -5.133393  1.671136     -1.898153       -2.374889   

   cons_conf_idx  euribor3m  nr_employed  y  ...  month_oct  month_sep  \
0       0.951267   0.773575     0.845170  0  ...      False      False   
1      -0.323542   0.230456     0.398115  0  ...      False      False   
2       0.151810  -1.667578    -2.428157  1  ...      False      False   
3      -1.425496  -1.277824    -0.940281  0  ...      False      False   
4       1.966794  -1.586859    -1.257233  1  ...      False      False   

   day_of_week_fri  day_of_week_mon  day_of_week_thu  day_of_week_tue  \
0            False            F

In [13]:
from sklearn.model_selection import train_test_split

# Separar las características y la variable objetivo del conjunto de entrenamiento
X_train = dataframeTrain.drop(columns=['y'])  # Características
y_train = dataframeTrain['y']  # Variable objetivo

# Dividir el conjunto de datos de entrenamiento en subconjuntos de entrenamiento y validación
X_train, X_val, y_train, y_val = train_test_split(X_train, y_train, test_size=0.2, random_state=42)

# Conjunto de datos de prueba
X_test = dataframeTest  # Características



# procedemos con el entrenamiento del modelo de Regresión Logística.



In [20]:
from sklearn.linear_model import LogisticRegression

# Inicializar el modelo de Regresión Logística
logistic_model = LogisticRegression(max_iter=110000)  # Aumentar el número máximo de iteraciones

# Entrenar el modelo
logistic_model.fit(X_train, y_train)



LogisticRegression(max_iter=110000)

In [21]:
from sklearn.metrics import classification_report

# Realizar predicciones en el conjunto de validación
y_pred_logistic_val = logistic_model.predict(X_val)

# Evaluar el rendimiento del modelo en el conjunto de validación
print("Classification Report for Logistic Regression Model (Validation Set):")
print(classification_report(y_val, y_pred_logistic_val))





Classification Report for Logistic Regression Model (Validation Set):
              precision    recall  f1-score   support

           0       0.91      0.98      0.94      7295
           1       0.65      0.24      0.35       943

    accuracy                           0.90      8238
   macro avg       0.78      0.61      0.65      8238
weighted avg       0.88      0.90      0.88      8238



In [24]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.feature_selection import SelectFromModel
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.pipeline import Pipeline
from sklearn.metrics import classification_report
from sklearn.linear_model import LogisticRegression

# Cargar los datos
dataframeTrain = pd.read_csv("credit-train.csv")
dataframeTest = pd.read_csv("credit-test.csv")

# Preprocesamiento de los datos
def preprocess_data(dataframe):
    # Convertir variables categóricas en variables dummy
    dataframe = pd.get_dummies(dataframe, columns=['job', 'marital', 'education', 'default', 'housing', 'loan', 'contact', 'month', 'day_of_week', 'poutcome'])
    
    # Normalizar las características numéricas
    scaler = StandardScaler()
    numeric_features = ['age', 'duration', 'campaign', 'pdays', 'previous', 'emp_var_rate', 'cons_price_idx', 'cons_conf_idx', 'euribor3m', 'nr_employed']
    dataframe[numeric_features] = scaler.fit_transform(dataframe[numeric_features])
    
    return dataframe

# Dividir el conjunto de datos en características y variable objetivo
X_train = dataframeTrain.drop(columns=['y'])
y_train = dataframeTrain['y']

# Preprocesamiento de los datos de entrenamiento y prueba
X_train = preprocess_data(X_train)
X_test = preprocess_data(dataframeTest)

# Dividir el conjunto de datos de entrenamiento en subconjuntos de entrenamiento y validación
X_train, X_val, y_train, y_val = train_test_split(X_train, y_train, test_size=0.2, random_state=42)

# Crear un pipeline para la selección de características y el modelo
pipeline = Pipeline([
    ('feature_selection', SelectFromModel(RandomForestClassifier())),
    ('classification', LogisticRegression(max_iter=110000))
])

# Definir los parámetros para la búsqueda de hiperparámetros
parameters = {
    'feature_selection__estimator__n_estimators': [50, 100, 200],
    'feature_selection__threshold': ['mean', 'median', '1.25*mean']
}

# Realizar la búsqueda de hiperparámetros utilizando validación cruzada
grid_search = GridSearchCV(pipeline, parameters, cv=5, n_jobs=-1, scoring='f1')
grid_search.fit(X_train, y_train)

# Obtener el mejor modelo
best_model = grid_search.best_estimator_

# Realizar predicciones en el conjunto de validación
y_pred_val = best_model.predict(X_val)

# Evaluar el rendimiento del modelo en el conjunto de validación
print("Classification Report for Logistic Regression Model (Validation Set):")
print(classification_report(y_val, y_pred_val))

# Realizar predicciones en el conjunto de prueba
y_pred_test = best_model.predict(X_test)

# Guardar las predicciones en un archivo CSV
predictions_df = pd.DataFrame({'y_pred': y_pred_test})
predictions_df.to_csv('predictions.csv', index=False)


Classification Report for Logistic Regression Model (Validation Set):
              precision    recall  f1-score   support

           0       0.93      0.97      0.95      7295
           1       0.64      0.41      0.50       943

    accuracy                           0.91      8238
   macro avg       0.78      0.69      0.72      8238
weighted avg       0.89      0.91      0.90      8238



In [25]:
from sklearn.metrics import classification_report

# Realizar predicciones en el conjunto de prueba
y_pred = logistic_model.predict(X_test)

# Evaluar el rendimiento del modelo
print("Classification Report for Logistic Regression Model:")
print(classification_report(y_test, y_pred))


ValueError: The feature names should match those that were passed during fit.
Feature names unseen at fit time:
- duration


In [23]:
# Definir y_test utilizando la columna 'y' del conjunto de datos de prueba
y_test = dataframeTest['y']


KeyError: 'y'

In [26]:
print("Primeras filas del conjunto de entrenamiento:")
print(dataframeTrain.head())

print("\nPrimeras filas del conjunto de prueba:")
print(dataframeTest.head())


Primeras filas del conjunto de entrenamiento:
   age          job  marital          education  default housing loan  \
0   44  blue-collar  married           basic.4y  unknown     yes   no   
1   53   technician  married            unknown       no      no   no   
2   28   management   single  university.degree       no     yes   no   
3   39     services  married        high.school       no      no   no   
4   55      retired  married           basic.4y       no     yes   no   

    contact month day_of_week  ...  campaign  pdays  previous     poutcome  \
0  cellular   aug         thu  ...         1    999         0  nonexistent   
1  cellular   nov         fri  ...         1    999         0  nonexistent   
2  cellular   jun         thu  ...         3      6         2      success   
3  cellular   apr         fri  ...         2    999         0  nonexistent   
4  cellular   aug         fri  ...         1      3         1      success   

  emp_var_rate  cons_price_idx  cons_conf_idx 

In [27]:
from sklearn.metrics import classification_report

# Definir y_test
y_test = dataframeTest['y']

# Realizar predicciones en el conjunto de prueba
y_pred_logistic = logistic_model.predict(X_test)

# Evaluar el rendimiento del modelo de regresión logística
print("Classification Report for Logistic Regression Model:")
print(classification_report(y_test, y_pred_logistic))




KeyError: 'y'

In [28]:
from sklearn.metrics import classification_report

# Definir y_test utilizando la columna 'y' del conjunto de datos de prueba
y_test = dataframeTest['y']

# Realizar predicciones en el conjunto de prueba
y_pred_logistic = logistic_model.predict(X_test)

# Evaluar el rendimiento del modelo
print("Classification Report for Logistic Regression Model:")
print(classification_report(y_test, y_pred_logistic))


KeyError: 'y'

In [21]:
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report

# Cargar los datos
dataframeTrain = pd.read_csv("credit-train.csv")
dataframeTest = pd.read_csv("credit-test.csv")

# Crear variables dummy para las columnas categóricas
columns_to_dummy = ['job', 'marital', 'education', 'default', 'housing', 'loan', 'contact']
dataframeTrain = pd.get_dummies(dataframeTrain, columns=columns_to_dummy)
dataframeTest = pd.get_dummies(dataframeTest, columns=columns_to_dummy)

# Normalizar los datos numéricos
scaler = StandardScaler()
numeric_columns = ['age', 'campaign', 'pdays', 'previous', 'emp_var_rate', 
                   'cons_price_idx', 'cons_conf_idx', 'euribor3m', 'nr_employed']
dataframeTrain[numeric_columns] = scaler.fit_transform(dataframeTrain[numeric_columns])
dataframeTest[numeric_columns] = scaler.transform(dataframeTest[numeric_columns])

# Eliminar columnas no relevantes
columns_to_drop = ['duration']  # Columnas a eliminar
dataframeTrain.drop(columns=columns_to_drop, inplace=True)
dataframeTest.drop(columns=columns_to_drop, inplace=True)

# Dividir los datos en conjunto de entrenamiento, validación y prueba
X = dataframeTrain.drop(columns=['y'])
y = dataframeTrain['y']
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)

# Separar la columna 'y' del conjunto de datos de prueba
X_test = dataframeTest.drop(columns=['y'])
y_test = dataframeTest['y']


# Entrenar el modelo de regresión logística
logistic_model = LogisticRegression()
logistic_model.fit(X_train, y_train)

# Realizar predicciones en el conjunto de prueba
y_pred_logistic = logistic_model.predict(X_test)

# Evaluar el rendimiento del modelo
print("Classification Report for Logistic Regression Model:")
print(classification_report(y_test, y_pred_logistic))


KeyError: "['y'] not found in axis"

In [22]:
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report

# Cargar el conjunto de datos de entrenamiento
dataframeTrain = pd.read_csv("credit-train.csv")
# Cargar el conjunto de datos de prueba
dataframeTest = pd.read_csv("credit-test.csv")

# Crear variables dummy para las columnas categóricas
columns_to_dummy = ['job', 'marital', 'education', 'default', 'housing', 'loan', 'contact']
dataframeTrain = pd.get_dummies(dataframeTrain, columns=columns_to_dummy)
dataframeTest = pd.get_dummies(dataframeTest, columns=columns_to_dummy)

# Normalizar los datos numéricos
scaler = StandardScaler()
numeric_columns = ['age', 'campaign', 'pdays', 'previous', 'emp_var_rate', 
                   'cons_price_idx', 'cons_conf_idx', 'euribor3m', 'nr_employed']
dataframeTrain[numeric_columns] = scaler.fit_transform(dataframeTrain[numeric_columns])
dataframeTest[numeric_columns] = scaler.transform(dataframeTest[numeric_columns])

# Eliminar columnas no relevantes
columns_to_drop = ['duration']  # Columnas a eliminar
dataframeTrain.drop(columns=columns_to_drop, inplace=True)
dataframeTest.drop(columns=columns_to_drop, inplace=True)

# Dividir los datos en conjunto de entrenamiento, validación y prueba
X_train = dataframeTrain.drop(columns=['y'])
y_train = dataframeTrain['y']
X_test = dataframeTest.drop(columns=['y'])
y_test = dataframeTest['y']

# Entrenar el modelo de regresión logística
logistic_model = LogisticRegression(max_iter=1000)
logistic_model.fit(X_train, y_train)

# Realizar predicciones en el conjunto de prueba
y_pred_logistic = logistic_model.predict(X_test)

# Evaluar el rendimiento del modelo
print("Classification Report for Logistic Regression Model:")
print(classification_report(y_test, y_pred_logistic))


KeyError: "['y'] not found in axis"